granulate_utils/linux/elf.py

```diff
@@ -4,13 +4,41 @@
 #
 
 import hashlib
-from typing import Optional, cast
+from functools import wraps
+from typing import Callable, Optional, TypeVar, cast
 
+import psutil
 from elftools.elf.elffile import ELFError, ELFFile  # type: ignore
 from elftools.elf.sections import NoteSection  # type: ignore
+from typing_extensions import ParamSpec
 
 __all__ = ["ELFError"]
 
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+def raise_nosuchprocess(func: Callable[P, R]) -> Callable[P, R]:
+    @wraps(func)
+    def inner(*args: P.args, **kwargs: P.kwargs):
+        try:
+            return func(*args, **kwargs)
+        except FileNotFoundError as e:
+            # Check if filename is /proc/{pid}/*
+            if e.filename.startswith("/proc/"):
+                if e.filename.split("/")[2].isalnum():
+                    # Take pid from /proc/{pid}/*
+                    pid = int(e.filename.split("/")[2])
+                    # Check if number from /proc/{pid} is actually a pid number
+                    with open("/proc/sys/kernel/pid_max") as pid_max_file:
+                        pid_max = int(pid_max_file.read())
+                    if pid <= pid_max:
+                        # Check if pid is running
+                        if not psutil.pid_exists(pid):
```

Comment:
It is true though, that `pid_exists` is a better check than what we have in `read_proc_file` - we use `os.path.exists("/proc/pid")`, but this will return True also if `pid` is a `tid` of another process (e.g due to reuse). `pid_exists` uses different means to check, to avoid this confusion. Note @d3dave .

Reply:
Also, instead of `exists(/proc/pid)` should have just `kill(pid, 0)`

Reply:
`pid_exists` does extra checks because they claim `os.kill` is bad as well :dumb: